keras/utils/data_utils.py

```diff
@@ -282,3 +295,295 @@ def validate_file(fpath, file_hash, algorithm='auto', chunk_size=65535):
+    enqueuer.close()
+    ```
+
+    The `enqueuer.get()` should be an infinite stream of datas.
+
+    """
+
+    @abstractmethod
+    def is_running(self):
+        raise NotImplemented
+
+    @abstractmethod
+    def start(self, workers=1, max_q_size=10):
+        """Start the handler's workers.
+
+        # Arguments
+            workers: number of worker threads
+            max_q_size: queue size (when full, threads could block on put())
+        """
+        raise NotImplemented
+
+    @abstractmethod
+    def stop(self, timeout=None):
+        """Stop running threads and wait for them to exit, if necessary.
+
+        Should be called by the same thread which called start().
+
+        # Arguments
+            timeout: maximum time to wait on thread.join()
+        """
+        raise NotImplemented
+
+    @abstractmethod
+    def get(self):
+        """Create a generator to extract data from the queue.
+
+        #Returns
+            A generator
+        """
+        raise NotImplemented
+
+
+class OrderedEnqueuer(DatasetEnqueuer):
+    """Builds a Enqueuer from a Dataset.
+
+    Used in `fit_generator`, `evaluate_generator`, `predict_generator`.
+
+    # Arguments
+        dataset: A `keras.data.dataset.Dataset` object.
+        pickle_safe: use multiprocessing if True, otherwise threading
+        scheduling: Sequential querying of datas if 'sequential', random otherwise.
+    """
+
+    def __init__(self, dataset, pickle_safe=False, scheduling='sequential'):
+        self.dataset = dataset
+        self.pickle_safe = pickle_safe
+        self.scheduling = scheduling
+        self.workers = 0
+        self.executor = None
+        self.queue = None
+        self.run_thread = None
+        self.stop_signal = None
+
+    def is_running(self):
+        return self.stop_signal is not None and not self.stop_signal.is_set()
+
+    def start(self, workers=1, max_q_size=10):
+        """Start the handler's workers.
+
+        # Arguments
+            workers: number of worker threads
+            max_q_size: queue size (when full, workers could block on put())
+        """
+        if self.pickle_safe:
+            self.executor = multiprocessing.Pool(workers)
+        else:
+            self.executor = ThreadPool(workers)
+        self.queue = queue.Queue(max_q_size)
+        self.stop_signal = threading.Event()
+        self.run_thread = threading.Thread(target=self._run)
+        self.run_thread.daemon = True
+        self.run_thread.start()
+
+    def _run(self):
+        """ Function to submit request to the executor and queue the `Future` objects."""
+        indexes = list(range(len(self.dataset)))
+        while True:
+            if self.scheduling is not 'sequential':
+                random.shuffle(indexes)
+            for i in indexes:
+                if self.stop_signal.is_set():
+                    return
+                self.queue.put(self.executor.apply_async(get_index, (self.dataset, i)), block=True)
```

Comment:
Most errors should be handled by the user in their dataset implementation. 

That said perhaps we should define an exception for users to raise which handles an index not being present by simply continuing? Or alternatively should there be a callback?